src/jdk.incubator.code/share/classes/jdk/incubator/code/internal/CodeModelToAST.java

```diff
@@ -0,0 +1,267 @@
+    }
+
+    private Type typeElementToType(TypeElement te) {
+        JavaType jt = (JavaType) te;
+        return switch (jt) {
+            case ClassType ct when ct.hasTypeArguments() -> {
+                ClassType enclosingType = ct.enclosingType().orElse(null);
+                ListBuffer<Type> typeArgs = new ListBuffer<>();
+                for (JavaType typeArgument : ct.typeArguments()) {
+                    typeArgs.add(typeElementToType(typeArgument));
+                }
+                yield new Type.ClassType(typeElementToType(enclosingType), typeArgs.toList(),
+                        typeElementToType(ct.rawType()).tsym);
+            }
+            case ArrayType at -> {
+                yield new Type.ArrayType(typeElementToType(at.componentType()), syms.arrayClass);
+            }
+            case null -> Type.noType;
+            default -> {
+                if (!jtToType.containsKey(te)) {
+                    throw new IllegalStateException("JavaType -> Type not found for: " + te.externalize().toString());
+                }
+                yield jtToType.get(te);
+            }
+        };
+    }
+
+    private JCTree invokeOpToJCMethodInvocation(CoreOp.InvokeOp invokeOp) {
+        Method method;
+        try {
+            method = invokeOp.invokeDescriptor().resolveToDirectMethod(MethodHandles.lookup());
+        } catch (ReflectiveOperationException e) {
+            throw new RuntimeException(e);
+        }
+        long flags = method.getModifiers();
+        var name = names.fromString(invokeOp.invokeDescriptor().name());
+        Value receiver = invokeOp.invokeKind() == CoreOp.InvokeOp.InvokeKind.INSTANCE ? invokeOp.operands().get(0) : null;
+        List<Value> arguments = invokeOp.operands().stream().skip(receiver == null ? 0 : 1).collect(List.collector());
+        var paramTypes = invokeOp.invokeDescriptor().type().parameterTypes().stream().map(this::typeElementToType)
+                .collect(List.collector());
+        var restype = typeElementToType(invokeOp.resultType());
+        var type = new Type.MethodType(paramTypes, restype, List.nil(), syms.methodClass);
+        var methodSym = new Symbol.MethodSymbol(flags, name, type,
+                typeElementToType(invokeOp.invokeDescriptor().refType()).tsym);
+        var meth = receiver == null ? treeMaker.Ident(methodSym) : treeMaker.Select((JCTree.JCExpression) valueToTree.get(receiver), methodSym);
+        var args = new ListBuffer<JCTree.JCExpression>();
+        for (Value operand : arguments) {
+            args.add((JCTree.JCExpression) valueToTree.get(operand));
+        }
+        var methodInvocation = treeMaker.App(meth, args.toList());
+        if (invokeOp.isVarArgs()) {
+            var lastParam = invokeOp.invokeDescriptor().type().parameterTypes().getLast();
+            Assert.check(lastParam instanceof ArrayType);
+            methodInvocation.varargsElement = typeElementToType(((ArrayType) lastParam).componentType());
```

Comment:
Same problem as described above -- here we're copying the vararg array at the declaration site into a type at the use-site -- for something like `List::of` this won't give the result you expect.

Reply:
Now -- while all the points I made above apply (e.g. if you were to try and type-check the generated AST using `Attr` you will get several errors), the saving grace here is that you are sending this tree into javac's backend anyway. And the backend plays a lot looser with types, only inserting casts where absolutely needed. Since you are using the op result type on the `MethodType` instance you are generating, I believe that should be enough for the backend to at least insert 90% of the type conversions that might be required because of erasure. So, in practice, even if incorrect, the code above might work fine.